lib/elixir_sense/core/parser.ex

```diff
@@ -19,40 +19,66 @@ defmodule ElixirSense.Core.Parser do
       {:ok, ast, modified_source} ->
         acc = MetadataBuilder.build(ast)
         if Map.has_key?(acc.lines_to_env, cursor_line_number) or !try_to_fix_line_not_found  do
-          create_metadata(source, acc)
+          create_metadata(source, {:ok, acc})
         else
           # IO.puts :stderr, "LINE NOT FOUND"
-          # match here as fix_line_not_found should not introduce syntax errors
-          {:ok, ast, _modified_source} = modified_source
-          |> fix_line_not_found(cursor_line_number)
-          |> string_to_ast(0, cursor_line_number)
-
-          acc = MetadataBuilder.build(ast)
-          # a line has been inserted, insert a fake lines_to_env
-          fixed_Lines_to_env = for {line, env} <- acc.lines_to_env, into: %{}, do: (
-            if line > cursor_line_number do
-              {line - 1, env}
-            else
-              {line , env}
-            end
-          )
-          acc = %ElixirSense.Core.State{acc | lines_to_env: fixed_Lines_to_env}
-
-          create_metadata(source, acc)
+          result = case try_fix_line_not_found_by_inserting_marker(modified_source, cursor_line_number) do
+            {:ok, acc} -> {:ok, acc}
+            _ -> try_fix_line_not_found_by_taking_previous_line(acc, cursor_line_number)
+          end
+
+          create_metadata(source, result)
         end
-      {:error, error} ->
+      {:error, reason} = error ->
         # IO.puts :stderr, "CAN'T FIX IT"
-        # IO.inspect :stderr, error, []
-        %Metadata{
-          source: source,
-          error: error
-        }
+        # IO.inspect :stderr, reason, []
+        create_metadata(source, error)
+    end
+  end
+
+  defp try_fix_line_not_found_by_taking_previous_line(acc, cursor_line_number) do
+    with {line, env} <- acc.lines_to_env
+    |> Enum.max_by(fn
```

Comment:
I would extract this whole function call to a `max_line/1` function and hide away all the complexity. Additionally, if you ever get `0` in the match on L43, it's not clear the `nil` in L50 covers all the cases, to me.

Reply:
I refactored this function and decided to return fallback empty metadata instead of returning an error